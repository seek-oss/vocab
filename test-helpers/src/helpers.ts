/* eslint-disable no-console */
import path from 'path';

import webpack from 'webpack';
import WDS from 'webpack-dev-server';
import { createServer, preview, build, loadConfigFromFile } from 'vite';
import waitOn from 'wait-on';

import { spawn } from 'child_process';

import { compile, resolveConfigSync } from '@vocab/core';

type Bundler = 'webpack' | 'vite';

interface Options {
  disableVocabPlugin?: boolean;
  bundler?: Bundler;
}

export interface TestServer {
  url: string;
  close: () => void;
}

let portCounter = 10001;

// TODO: Make `resolveConfigSync` accept a custom working directory so we don't need this map

const configExtensionByFixtureName = {
  direct: 'js',
  phrase: 'js',
  server: 'js',
  simple: 'cjs',
  vite: 'cjs',
  'translation-types': 'js',
} as const satisfies Record<string, 'js' | 'cjs'>;

export type FixtureName = keyof typeof configExtensionByFixtureName;

export const compileFixtureTranslations = async (fixtureName: FixtureName) => {
  const configExtension = configExtensionByFixtureName[fixtureName];

  const config = resolveConfigSync(
    require.resolve(
      `@vocab-fixtures/${fixtureName}/vocab.config.${configExtension}`,
    ),
  );

  if (!config) {
    throw new Error(
      `Can't resolve "@vocab-fixtures/${fixtureName}" vocab config`,
    );
  }

  await compile({ watch: false }, config);
};

export const runServerFixture = (
  fixtureName: FixtureName,
): Promise<TestServer> =>
  new Promise(async (resolve) => {
    await compileFixtureTranslations(fixtureName);

    const port = portCounter++;

    const getConfig = (
      await import(`@vocab-fixtures/${fixtureName}/webpack.config.js`)
    ).default;

    const config = getConfig();
    const compiler = webpack(config);

    compiler.hooks.done.tap('vocab-test-helper', async () => {
      const cwd = path.dirname(
        require.resolve(`@vocab-fixtures/${fixtureName}/package.json`),
      );
      const childProcess = spawn('node', ['./dist-server/server.js'], {
        env: { ...process.env, SERVER_PORT: port.toString() },
        stdio: 'inherit',
        cwd,
      });
      await waitOn({ resources: [`http://localhost:${port}`] });

      resolve({
        url: `http://localhost:${port}`,
        close: () => {
          childProcess.kill();
        },
      });
    });
    compiler.run(() => {
      // Run webpack build
    });
  });

type StartFixtureFunction = (
  fixtureName: FixtureName,
  options?: Options,
) => Promise<TestServer>;

export const startWebpackFixture: StartFixtureFunction = (
  fixtureName: FixtureName,
  options: Options = {},
): Promise<TestServer> =>
  new Promise(async (resolve) => {
    await compileFixtureTranslations(fixtureName);

    const getConfig = (
      await import(`@vocab-fixtures/${fixtureName}/webpack.config.js`)
    ).default;

    const config = getConfig(options);
    const compiler = webpack(config);

    const port = portCounter++;
    const server = new WDS({ port }, compiler);

    compiler.hooks.done.tap('vocab-test-helper', () => {
      resolve({
        url: `http://localhost:${port}`,
        close: () => server.stopCallback(),
      });
    });

    server.startCallback(() => {
      console.log(`Running fixture ${fixtureName}`);
    });
  });

export const startViteFixture: StartFixtureFunction = async (
  fixtureName: FixtureName,
  options: Omit<Options, 'bundler'> = {},
) => {
  await compileFixtureTranslations(fixtureName);

  const root = path.dirname(
    require.resolve(`@vocab-fixtures/${fixtureName}/package.json`),
  );

  const config = loadConfigFromFile(
    {
      command: 'serve',
      mode: 'development',
    },
    undefined,
    root,
  );

  const port = portCounter++;
  const server = await createServer({
    ...config,
    ...(options.disableVocabPlugin ? { plugins: [] } : {}),
    root: path.dirname(
      require.resolve(`@vocab-fixtures/${fixtureName}/package.json`),
    ),
    server: {
      port,
    },
  });

  const devServer = await server.listen();

  console.log(`Running fixture ${fixtureName}`);

  return {
    url: `http://localhost:${devServer.config.server.port}`,
    close: async () => devServer.close(),
  };
};

export const previewViteFixture: StartFixtureFunction = async (
  fixtureName: FixtureName,
  options: Options = {},
) => {
  try {
    await compileFixtureTranslations(fixtureName);

    const root = path.dirname(
      require.resolve(`@vocab-fixtures/${fixtureName}/package.json`),
    );

    const port = portCounter++;
    await build({
      ...loadConfigFromFile(
        {
          command: 'build',
          mode: 'development',
        },
        undefined,
        root,
      ),
      ...(options.disableVocabPlugin ? { plugins: [] } : {}),
      root,
    });

    const server = await preview({
      ...loadConfigFromFile(
        {
          command: 'serve',
          isPreview: true,
          mode: 'development',
        },
        undefined,
        root,
      ),
      ...(options.disableVocabPlugin ? { plugins: [] } : {}),
      root,
      preview: {
        strictPort: true,
        port,
        open: false,
      },
    });

    console.log(`Running fixture ${fixtureName}`);

    return {
      url: server.resolvedUrls?.local[0] || `http://localhost:${port}`,
      close: async () => {
        await server.close();
      },
    };
  } catch (e) {
    console.error(e);
    throw e;
  }
};

const fixtureBundlerMap: Record<Bundler, StartFixtureFunction> = {
  webpack: startWebpackFixture,
  vite: startViteFixture,
};

export const startFixture = (
  fixtureName: FixtureName,
  options: Options = { bundler: 'webpack' },
): Promise<TestServer> =>
  fixtureBundlerMap[options.bundler ?? 'webpack'](fixtureName, options);

export const getAppSnapshot = async (
  url: string,
  warningFilter = () => true,
) => {
  const warnings: unknown[] = [];
  const errors: unknown[] = [];

  const page = await browser.newPage();

  page.on('console', (msg) => {
    if (msg.type() === 'warn') {
      warnings.filter(warningFilter).push(msg.text());
    }

    if (msg.type() === 'error') {
      errors.push(msg.text());
    }
  });

  const response = await page.goto(url, { waitUntil: 'networkidle0' });
  const sourceHtml = await response?.text();
  const clientRenderContent = await page.content();

  expect(warnings).toEqual([]);
  expect(errors).toEqual([]);

  return { sourceHtml, clientRenderContent };
};

export const getLanguageChunk = async ({
  serverUrl,
  language,
}: {
  serverUrl: string;
  language: string;
}) => {
  const response = await page.goto(`${serverUrl}/${language}-translations.js`);
  return await response?.text();
};
