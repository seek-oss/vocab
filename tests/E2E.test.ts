import { startFixture, TestServer } from './helpers';

describe('E2E', () => {
  let server: TestServer;

  beforeAll(async () => {
    server = await startFixture('fixture-simple');
  });

  beforeEach(async () => {
    await page.goto(server.url);
  });

  it('should default to english', async () => {
    const message = await page.waitForSelector('#message');

    await expect(message).toMatch('Hello world');
  });

  it('should switch to french', async () => {
    await page.click('button');

    const message = await page.waitForSelector('#message');

    await expect(message).toMatch('Bonjour monde');
  });

  afterAll(() => {
    server.close();
  });
});
