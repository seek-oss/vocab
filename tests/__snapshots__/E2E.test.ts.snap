// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`E2E Simple with plugin should return the expected en chunk 1`] = `
""use strict";
(self["webpackChunk_vocab_private_monorepo"] = self["webpackChunk_vocab_private_monorepo"] || []).push([["en-translations"],{

/***/ "./fixtures/simple/src/.vocab/translations.json-en-virtual.json!=!./packages/virtual-resource-loader/dist/virtual-resource-loader.cjs.js?source=eyJoZWxsbyI6IkhlbGxvIiwid29ybGQiOiJ3b3JsZCJ9!":
/*!****************************************************************************************************************************************************************************************************!*\\
  !*** ./fixtures/simple/src/.vocab/translations.json-en-virtual.json!=!./packages/virtual-resource-loader/dist/virtual-resource-loader.cjs.js?source=eyJoZWxsbyI6IkhlbGxvIiwid29ybGQiOiJ3b3JsZCJ9! ***!
  \\****************************************************************************************************************************************************************************************************/
/***/ ((module) => {

module.exports = /*#__PURE__*/JSON.parse('{"hello":"Hello","world":"world"}');

/***/ }),

/***/ "./fixtures/simple/src/client.vocab/translations.json-en-virtual.json!=!./packages/virtual-resource-loader/dist/virtual-resource-loader.cjs.js?source=eyJzcGVjaWFsQ2hhcmFjdGVycyAtICfigJjigJnigJzigJ1cIiFAIyQlXiYqKClfK1xcL2B%2BXFxcXCI6IuKAmOKAmeKAnOKAnSdcIiFAIyQlXiYqKClfK1xcL2B%2BXFxcXCIsInZvY2FiUHVibGlzaERhdGUiOiI8c3Ryb25nPlZvY2FiPC9zdHJvbmc%2BIHdhcyBwdWJsaXNoZWQgb24ge3B1Ymxpc2hEYXRlLCBkYXRlLCBzbWFsbH0ifQ%3D%3D!":
/*!************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************!*\\
  !*** ./fixtures/simple/src/client.vocab/translations.json-en-virtual.json!=!./packages/virtual-resource-loader/dist/virtual-resource-loader.cjs.js?source=eyJzcGVjaWFsQ2hhcmFjdGVycyAtICfigJjigJnigJzigJ1cIiFAIyQlXiYqKClfK1xcL2B%2BXFxcXCI6IuKAmOKAmeKAnOKAnSdcIiFAIyQlXiYqKClfK1xcL2B%2BXFxcXCIsInZvY2FiUHVibGlzaERhdGUiOiI8c3Ryb25nPlZvY2FiPC9zdHJvbmc%2BIHdhcyBwdWJsaXNoZWQgb24ge3B1Ymxpc2hEYXRlLCBkYXRlLCBzbWFsbH0ifQ%3D%3D! ***!
  \\************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************/
/***/ ((module) => {

module.exports = /*#__PURE__*/JSON.parse('{"specialCharacters - \\'‘’“”\\\\"!@#$%^&*()_+\\\\\\\\/\`~\\\\\\\\\\\\\\\\":"‘’“”\\'\\\\"!@#$%^&*()_+\\\\\\\\/\`~\\\\\\\\\\\\\\\\","vocabPublishDate":"<strong>Vocab</strong> was published on {publishDate, date, small}"}');

/***/ })

}]);"
`;

exports[`E2E Simple with plugin should return the expected fr chunk 1`] = `
""use strict";
(self["webpackChunk_vocab_private_monorepo"] = self["webpackChunk_vocab_private_monorepo"] || []).push([["fr-translations"],{

/***/ "./fixtures/simple/src/.vocab/translations.json-fr-virtual.json!=!./packages/virtual-resource-loader/dist/virtual-resource-loader.cjs.js?source=eyJoZWxsbyI6IkJvbmpvdXIiLCJ3b3JsZCI6Im1vbmRlIn0%3D!":
/*!**********************************************************************************************************************************************************************************************************!*\\
  !*** ./fixtures/simple/src/.vocab/translations.json-fr-virtual.json!=!./packages/virtual-resource-loader/dist/virtual-resource-loader.cjs.js?source=eyJoZWxsbyI6IkJvbmpvdXIiLCJ3b3JsZCI6Im1vbmRlIn0%3D! ***!
  \\**********************************************************************************************************************************************************************************************************/
/***/ ((module) => {

module.exports = /*#__PURE__*/JSON.parse('{"hello":"Bonjour","world":"monde"}');

/***/ }),

/***/ "./fixtures/simple/src/client.vocab/translations.json-fr-virtual.json!=!./packages/virtual-resource-loader/dist/virtual-resource-loader.cjs.js?source=eyJzcGVjaWFsQ2hhcmFjdGVycyAtICfigJjigJnigJzigJ1cIiFAIyQlXiYqKClfK1xcL2B%2BXFxcXCI6IuKAmOKAmeKAnOKAnSdcIiFAIyQlXiYqKClfK1xcL2B%2BXFxcXCIsInZvY2FiUHVibGlzaERhdGUiOiJWb2NhYiBhIMOpdMOpIHB1Ymxpw6kgbGUge3B1Ymxpc2hEYXRlLCBkYXRlLCBtZWRpdW19In0%3D!":
/*!************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************!*\\
  !*** ./fixtures/simple/src/client.vocab/translations.json-fr-virtual.json!=!./packages/virtual-resource-loader/dist/virtual-resource-loader.cjs.js?source=eyJzcGVjaWFsQ2hhcmFjdGVycyAtICfigJjigJnigJzigJ1cIiFAIyQlXiYqKClfK1xcL2B%2BXFxcXCI6IuKAmOKAmeKAnOKAnSdcIiFAIyQlXiYqKClfK1xcL2B%2BXFxcXCIsInZvY2FiUHVibGlzaERhdGUiOiJWb2NhYiBhIMOpdMOpIHB1Ymxpw6kgbGUge3B1Ymxpc2hEYXRlLCBkYXRlLCBtZWRpdW19In0%3D! ***!
  \\************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************/
/***/ ((module) => {

module.exports = /*#__PURE__*/JSON.parse('{"specialCharacters - \\'‘’“”\\\\"!@#$%^&*()_+\\\\\\\\/\`~\\\\\\\\\\\\\\\\":"‘’“”\\'\\\\"!@#$%^&*()_+\\\\\\\\/\`~\\\\\\\\\\\\\\\\","vocabPublishDate":"Vocab a été publié le {publishDate, date, medium}"}');

/***/ })

}]);"
`;

exports[`E2E Simple with plugin should return the expected pseudo chunk 1`] = `
""use strict";
(self["webpackChunk_vocab_private_monorepo"] = self["webpackChunk_vocab_private_monorepo"] || []).push([["pseudo-translations"],{

/***/ "./fixtures/simple/src/.vocab/translations.json-pseudo-virtual.json!=!./packages/virtual-resource-loader/dist/virtual-resource-loader.cjs.js?source=eyJoZWxsbyI6IlvhuKjhur3hur3Gmsaaw7bDtl0iLCJ3b3JsZCI6IlvFtcO2w7bDtsWZxprGjF0ifQ%3D%3D!":
/*!************************************************************************************************************************************************************************************************************************************************!*\\
  !*** ./fixtures/simple/src/.vocab/translations.json-pseudo-virtual.json!=!./packages/virtual-resource-loader/dist/virtual-resource-loader.cjs.js?source=eyJoZWxsbyI6IlvhuKjhur3hur3Gmsaaw7bDtl0iLCJ3b3JsZCI6IlvFtcO2w7bDtsWZxprGjF0ifQ%3D%3D! ***!
  \\************************************************************************************************************************************************************************************************************************************************/
/***/ ((module) => {

module.exports = /*#__PURE__*/JSON.parse('{"hello":"[Ḩẽẽƚƚöö]","world":"[ŵöööřƚƌ]"}');

/***/ }),

/***/ "./fixtures/simple/src/client.vocab/translations.json-pseudo-virtual.json!=!./packages/virtual-resource-loader/dist/virtual-resource-loader.cjs.js?source=eyJzcGVjaWFsQ2hhcmFjdGVycyAtICfigJjigJnigJzigJ1cIiFAIyQlXiYqKClfK1xcL2B%2BXFxcXCI6IlvigJjigJnigJzigJ0nXCIhQCMkJV4mKigpXytcXC9gflxcXFxdIiwidm9jYWJQdWJsaXNoRGF0ZSI6Ils8c3Ryb25nPuG5vMO2w7bDp8SDxIPDnzwvc3Ryb25nPiDFtcSDxIPEg8WhIMalx5rHmseaw5%2FGmsOsw6zDrMWh4bip4bq94bq94bq9xowgw7bDtsO24LiB4Li14LmJIHtwdWJsaXNoRGF0ZSwgZGF0ZSwgc21hbGx9XSJ9!":
/*!**************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************!*\\
  !*** ./fixtures/simple/src/client.vocab/translations.json-pseudo-virtual.json!=!./packages/virtual-resource-loader/dist/virtual-resource-loader.cjs.js?source=eyJzcGVjaWFsQ2hhcmFjdGVycyAtICfigJjigJnigJzigJ1cIiFAIyQlXiYqKClfK1xcL2B%2BXFxcXCI6IlvigJjigJnigJzigJ0nXCIhQCMkJV4mKigpXytcXC9gflxcXFxdIiwidm9jYWJQdWJsaXNoRGF0ZSI6Ils8c3Ryb25nPuG5vMO2w7bDp8SDxIPDnzwvc3Ryb25nPiDFtcSDxIPEg8WhIMalx5rHmseaw5%2FGmsOsw6zDrMWh4bip4bq94bq94bq9xowgw7bDtsO24LiB4Li14LmJIHtwdWJsaXNoRGF0ZSwgZGF0ZSwgc21hbGx9XSJ9! ***!
  \\**************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************/
/***/ ((module) => {

module.exports = /*#__PURE__*/JSON.parse('{"specialCharacters - \\'‘’“”\\\\"!@#$%^&*()_+\\\\\\\\/\`~\\\\\\\\\\\\\\\\":"[‘’“”\\'\\\\"!@#$%^&*()_+\\\\\\\\/\`~\\\\\\\\\\\\\\\\]","vocabPublishDate":"[<strong>Ṽööçăăß</strong> ŵăăăš ƥǚǚǚßƚìììšḩẽẽẽƌ öööกี้ {publishDate, date, small}]"}');

/***/ })

}]);"
`;

exports[`E2E vite with plugin should return the expected en chunk 1`] = `
"const o="Hello",t="world",l={hello:o,world:t},r=Object.freeze(Object.defineProperty({__proto__:null,default:l,hello:o,world:t},Symbol.toStringTag,{value:"Module"})),e="<strong>Vocab</strong> was published on {publishDate, date, small}",a={"specialCharacters - '‘’“”\\"!@#$%^&*()_+\\\\/\`~\\\\\\\\":"‘’“”'\\"!@#$%^&*()_+\\\\/\`~\\\\\\\\",vocabPublishDate:e},n=Object.freeze(Object.defineProperty({__proto__:null,default:a,vocabPublishDate:e},Symbol.toStringTag,{value:"Module"}));export{r as _,n as a};
"
`;

exports[`E2E vite with plugin should return the expected fr chunk 1`] = `
"const e="Bonjour",o="monde",a={hello:e,world:o},r=Object.freeze(Object.defineProperty({__proto__:null,default:a,hello:e,world:o},Symbol.toStringTag,{value:"Module"})),t="Vocab a été publié le {publishDate, date, medium}",l={"specialCharacters - '‘’“”\\"!@#$%^&*()_+\\\\/\`~\\\\\\\\":"‘’“”'\\"!@#$%^&*()_+\\\\/\`~\\\\\\\\",vocabPublishDate:t},c=Object.freeze(Object.defineProperty({__proto__:null,default:l,vocabPublishDate:t},Symbol.toStringTag,{value:"Module"}));export{r as _,c as a};
"
`;

exports[`E2E vite with plugin should return the expected pseudo chunk 1`] = `
"const t="[Ḩẽẽƚƚöö]",o="[ŵöööřƚƌ]",s={hello:t,world:o},a=Object.freeze(Object.defineProperty({__proto__:null,default:s,hello:t,world:o},Symbol.toStringTag,{value:"Module"})),e="[<strong>Ṽööçăăß</strong> ŵăăăš ƥǚǚǚßƚìììšḩẽẽẽƌ öööกี้ {publishDate, date, small}]",b={"specialCharacters - '‘’“”\\"!@#$%^&*()_+\\\\/\`~\\\\\\\\":"[‘’“”'\\"!@#$%^&*()_+\\\\/\`~\\\\\\\\]",vocabPublishDate:e},l=Object.freeze(Object.defineProperty({__proto__:null,default:b,vocabPublishDate:e},Symbol.toStringTag,{value:"Module"}));export{l as G,a as _};
"
`;
